docs/src/explanations/optional-groups.md

```diff
@@ -0,0 +1,177 @@
+definition may use any Chisel or Scala value visible to its Scala lexical scope.
+
+The following example defines four optional groups inside module `Foo` and
+declares wires which are connected to values captured from visible lexical
+scope:
+
+```scala mdoc:silent
+import chisel3._
+import chisel3.group
+
+class Foo extends RawModule {
+  val port = IO(Input(Bool()))
+
+  group(A) {
+    val a = WireInit(port)
+    group(A.B) {
+      val b = WireInit(a)
+      group(A.B.C) {
+        val c = WireInit(b)
+      }
+    }
+    group(A.D) {
+      val d = WireInit(port ^ a)
+    }
+  }
+}
+```
+
+## Conventions
+
+Currently, there is only one supported convetion, `Bind`.  This will cause
+optional group definitions to be lowered to Verilog modules that are
+instantiated via the SystemVerilog `bind` mechanism.  The lowering to Verilog of
+nested groups avoids illegal nested usage of `bind`.
+
+More conventions may be supported in the future.
+
+## Examples
+
+### Design Verification Example
+
+Consider a use case where a design or design verification engineer would like to
+add some asserts and debug prints to a module.  The logic necessary for the
+asserts and debug prints requires additional computation.  All of this code
+should not be included in the final Verilog.  The engineer can use three
+optional groups to do this.
+
+There are three optional groups that emerge from this example:
+
+1. A common `Verification` group
+1. An `Assert` group nested under the `Verification` group
+1. A `Debug` group also nested under the `Verification` group
+
+The `Verification` group can be used to store common logic used by both the
+`Assert` and `Debug` groups.  The latter two groups allow for separation of,
+respectively, assertions from prints.
+
+One way to write this in Scala is the following:
+
+```scala mdoc:reset:silent
+import chisel3._
+import chisel3.group.{Convention, Declaration}
+
+// All groups are declared here.  The Assert and Debug groups are nested under
+// the Verification group.
+object Verification extends group.Declaration(group.Convention.Bind) {
+  object Assert extends group.Declaration(group.Convention.Bind)
+  object Debug extends group.Declaration(group.Convention.Bind)
+}
+
+class Foo extends Module {
+  val a = IO(Input(UInt(32.W)))
+  val b = IO(Output(UInt(32.W)))
+
+  b := a +% 1.U
+
+  // This defines the `Verification` group inside Foo.
+  group(Verification) {
+
+    // Some common logic added here.  The input port `a` is "captured" and
+    // used here.
+    val a_d0 = RegNext(a)
+
+    // This defines the `Assert` group.
+    group(Verification.Assert) {
+      chisel3.assert(a >= a_d0, "a must always increment")
+    }
+
+    // This defines the `Debug` group.
+    group(Verification.Debug) {
+      printf("a: %x, a_d0: %x", a, a_d0)
+    }
+  }
+
+}
+
+```
+
+After compilation, this will produce four Verilog files (one for `Foo` and one
+for each optional group definition in module `Foo`):
+
+1. `Foo`
```

Comment:
are these supposed to have the .sv extension? you mentioned these are files?

Reply:
Good catch. It should say "four Verilog modules". I'll change this.

Whether or not these are actual files isn't part of the ABI. Only the names and existence of the bind files are part of the ABI.